README.md
# LED_Display_arduino


Overview
--------------------------------------------
* Name. LED_Display_arduino
* Title. Arduino Based LED Display unit.
* Description. An Arduino Based LED Display unit with 15 LEDs and 20 lighting modes. 
  A Potentiometer selects mode and a push button activates selected mode.
* Author. Gavin Lyons
* URL. https://github.com/gavinlyonsrepo/LED_Display_arduino

Library
--------------------------
* Custom library written by author, Display_LED_library available at repository URL 
* Version 1.0.1 
* Display_LED_lib.h // to drive the LED arrays

Software Used
------------------
Arduino Version 1.8.5

Eagle Version 9.1.3

Files required
----------------------
* Main file. arduino_LED_Display.ino
* Custom library header. Display_LED_lib.h
* Custom library file. Display_LED_lib.cpp

Parts List
------------------------------
The schematic drawn in eagle is in docs folder. 

You will need following parts.

>
> ATmega328p micro-controller  as per schematic or Arduino uno/nano module.
>
> Resistors. 15 x 120ohms , 220ohms, 660ohms
>
> LEDS. 15, 5 red , 5 amber , 5 green 
>
> one 10 k potentiometer
>
> one pushbutton
>
> one 16 Mhz crystal 
>
> One Lm317 voltage regulator 
>
> Capacitors. two x 22pF, two x 100nF , 1uF
>

Features
----------------------
The user selects the mode via the potentiometer whose ADC range 0->5 volts (10 bit 1023) are mapped to 1-20.
For each of the twenty modes. The user presses the push button to activate the mode. This starts an Interrupt service routine which reads pot value and changes the mode.  The push button is hardware de-bounced by a capacitor.

There are 15 LEDs, five red, five amber, five green.

Most of the code is in a custom library developed for this project by author, called "Display_LED_lib".
The library is in the repository. 

The modes are 
1. All Leds off
2. All Leds on
3. All red Leds on
4. All amber Leds on
5. All green Leds on
6. knight rider_1  
7. knight rider_2
8. knight rider_3 
9. knight rider_4 
10. Traffic light sequence
11.  All LEDS Flash on and off
12. flash red-amber-green in rapid  sequence
13. Fade in and out the 6 Leds are connected to pins with PWM capability 
14. Random 1 led turn on and off
15. Random 3 led turn on and off
16. Random Multiple LEDs from each array 
17. Lone red blink
18. Lone amber blink
19. Lone green blink
20. all Modes combined repeat

LED pin layout

| Pin Uno | Pin Atmega  | PWM  | Colour |   
| --- | --- | --- | --- |
|  D3 |  PD3 | Yes  | red  |   
|  D4  | PD4  | no  |  red |   
|  D5 |  PD5 |  Yes | red  |   
| D6  |  PD6 |  Yes |  red |   
|  D7 |  PD7 | no  | red  |   
| D8  | PB0  |  no |  amber |   
| D9  | PB1  |   Yes|  amber  |   
| D10  | PB2  |  Yes | amber   |   
| D11  | PB3  |  Yes | amber   |   
| D12  | PB4  |  no | amber   |   
| D13  | PB5  |  no |  green |   
|  A0 | PC0  |  no |  green |   
|  A1 | PC1  |  no |  green |   
|  A2 | PC2  |  no | green  |   
| A3  | PC3  |  no |  green |   

![schematic image ](https://github.com/gavinlyonsrepo/LED_Display_arduino/blob/master/docs/eagle/led_display.png)
 
Copyright
---------
Copyright (C) 2018 Gavin Lyons 
see LICENSE.md in documentation section 
for more details



docs/CHANGELOG.md
# Changelog

* version 1.0-1 August 2018
	first release


	

lib/Display_LED_lib.h
// url:https://github.com/gavinlyonsrepo/LED_Display_arduino

#ifndef tl
#define tl

#if (ARDUINO >=100)
  #include "Arduino.h"
#else
  #include "WProgram.h"
#endif

class Display_LED_lib {
  public:
	// Constructor 
	Display_LED_lib(bool displayMsg=false);

	// Methods
	void begin(int baudRate=9600);

	void Ledmode_alloff();
	void Ledmode_allon();
    void Ledmode_redon();
	void Ledmode_amberon();
	void Ledmode_greenon();
	void Ledmode_knightrider();
	void Ledmode_knightrider_2();
	void Ledmode_knightrider_3();
	void Ledmode_knightrider_4();
	void Ledmode_trafficlight();
	void Ledmode_allblink();
	void Ledmode_trafficlight_fast();
	void Ledmode_pwm();
	void Ledmode_random();
	void Ledmode_random_2();
	void Ledmode_random_3();
	void Ledmode_redone();
	void Ledmode_amberone();
	void Ledmode_greenone();
	void Ledmode_all();
	void Ledmode_setup();
	void mydelay(int)


  private:
	bool _msg;
	int _period_100;
	int _period_250;
	int _period_500;
	int _period_750;
	int _period_1000;
	int _period_4000;
	int _period_7000;

	 //led arrays  of pin numbers to which LEDs are attached
	  int _pinCount = 5;
	  int _ledPinsRed[5] = {3, 4, 5, 6, 7};
	  int _ledPinsAmber[5] = {8, 9, 10, 11, 12};
	  int _ledPinsGreen[5] = {13, 14, 15, 16, 17};
	  int _ledPinsPwm[6] = {3, 5, 6, 9, 10, 11};

};
#endif


lib/Display_LED_lib.cpp

#include "Display_LED_lib.h"

Display_LED_lib::Display_LED_lib(bool displayMsg) {
  // Anything you need when instantiating your object goes here
  _msg = displayMsg;
}

// this is our 'begin' function
void Display_LED_lib::begin(int baudRate) {
  Serial.begin(baudRate);
  if (_msg) {
    Serial.println("TestLib constructor instantiated (created) successfully.");
  }
}


//setup
void Display_LED_lib::Ledmode_setup()
{
//LED Outputs
  for (int thisPin = 0; thisPin < _pinCount; thisPin++)
  {
    pinMode(_ledPinsRed[thisPin], OUTPUT);
    pinMode(_ledPinsAmber[thisPin], OUTPUT);
    pinMode(_ledPinsGreen[thisPin], OUTPUT);
  }
}

//1
void Display_LED_lib::Ledmode_alloff()
{
  for (int thisPin = 0; thisPin < _pinCount; thisPin++)
  {
    digitalWrite(_ledPinsRed[thisPin], LOW);
    digitalWrite(_ledPinsAmber[thisPin], LOW);
    digitalWrite(_ledPinsGreen[thisPin], LOW);
  }
}

//2
void Display_LED_lib::Ledmode_allon()
{
  for (int thisPin = 0; thisPin < _pinCount; thisPin++)
  {
    digitalWrite(_ledPinsRed[thisPin], HIGH);
    digitalWrite(_ledPinsAmber[thisPin], HIGH);
    digitalWrite(_ledPinsGreen[thisPin], HIGH);
  }
}

//3
void  Display_LED_lib::Ledmode_redon()
{
  for (int thisPin = 0; thisPin < _pinCount; thisPin++)
  {
    // turn the pin off:
    digitalWrite(_ledPinsRed[thisPin], HIGH);
    digitalWrite(_ledPinsAmber[thisPin], LOW);
    digitalWrite(_ledPinsGreen[thisPin], LOW);
  }
}

//4
void Display_LED_lib::Ledmode_amberon()
{
  for (int thisPin = 0; thisPin < _pinCount; thisPin++)
  {
    // turn the pin off:
    digitalWrite(_ledPinsRed[thisPin], LOW);
    digitalWrite(_ledPinsAmber[thisPin], HIGH);
    digitalWrite(_ledPinsGreen[thisPin], LOW);
  }
}

//5
void Display_LED_lib::Ledmode_greenon()
{
  for (int thisPin = 0; thisPin < _pinCount; thisPin++)
  {
    // turn the pin off:
    digitalWrite(_ledPinsRed[thisPin], LOW);
    digitalWrite(_ledPinsAmber[thisPin], LOW);
    digitalWrite(_ledPinsGreen[thisPin], HIGH);
  }
}

//6
void Display_LED_lib::Ledmode_knightrider()
{
  Ledmode_alloff();
  for (int thisPin = 0; thisPin < _pinCount; thisPin++)
  {
    digitalWrite(_ledPinsRed[thisPin], HIGH);
    mydelay(_period_250);
    // turn the pin off:
    digitalWrite(_ledPinsRed[thisPin], LOW);
  }
  for (int thisPin = 0; thisPin < _pinCount; thisPin++)
  {
    digitalWrite(_ledPinsAmber[thisPin], HIGH);
   mydelay(_period_250);
    // turn the pin off:
    digitalWrite(_ledPinsAmber[thisPin], LOW);
  }
  for (int thisPin = 0; thisPin < _pinCount; thisPin++)
  {
    digitalWrite(_ledPinsGreen[thisPin], HIGH);
    mydelay(_period_250);
    // turn the pin off:
    digitalWrite(_ledPinsGreen[thisPin], LOW);
  }

}

//7
void Display_LED_lib::Ledmode_knightrider_2()
{
  Ledmode_alloff();
  for (int thisPin = 0; thisPin < _pinCount; thisPin++)
  {
    digitalWrite(_ledPinsRed[thisPin], HIGH);
    mydelay(_period_250);
  }
  for (int thisPin = 0; thisPin < _pinCount; thisPin++)
  {
    digitalWrite(_ledPinsAmber[thisPin], HIGH);
    mydelay(_period_250);
  }
  for (int thisPin = 0; thisPin < _pinCount; thisPin++)
  {
    digitalWrite(_ledPinsGreen[thisPin], HIGH);
    mydelay(_period_250);
  }
}

//8
void Display_LED_lib::Ledmode_knightrider_3()
{
  Ledmode_alloff();
  for (int thisPin = 0; thisPin < _pinCount; thisPin++)
  {
    digitalWrite(_ledPinsRed[thisPin], HIGH);
    digitalWrite(_ledPinsAmber[thisPin], HIGH);
    digitalWrite(_ledPinsGreen[thisPin], HIGH);
    mydelay(_period_250);
  }
}

//9
void Display_LED_lib::Ledmode_knightrider_4()
{
  Ledmode_alloff();
  for (int thisPin = 0; thisPin < _pinCount; thisPin++)
  {
    digitalWrite(_ledPinsRed[thisPin], HIGH);
    digitalWrite(_ledPinsAmber[thisPin], HIGH);
    digitalWrite(_ledPinsGreen[thisPin], HIGH);
    mydelay(_period_250);
    digitalWrite(_ledPinsRed[thisPin], LOW);
    digitalWrite(_ledPinsAmber[thisPin], LOW);
    digitalWrite(_ledPinsGreen[thisPin], LOW);
  }
}

//10
void Display_LED_lib::Ledmode_trafficlight()
{
  Ledmode_alloff();
  Ledmode_greenon();
  mydelay(_period_7000);
  Ledmode_amberon();
  mydelay(_period_4000);
  Ledmode_redon();
  mydelay(_period_7000);
}

//11
void Display_LED_lib::Ledmode_allblink()
{
  Ledmode_allon();
  mydelay(_period_1000);
  Ledmode_alloff();
   mydelay(_period_1000);
}

//12
void Display_LED_lib::Ledmode_trafficlight_fast()
{
  Ledmode_alloff();
  Ledmode_greenon();
   mydelay(_period_750);
  Ledmode_amberon();
   mydelay(_period_750);
  Ledmode_redon();
  mydelay(_period_750);
}

//13
void Display_LED_lib::Ledmode_pwm()
{
  Ledmode_alloff();
  // fade in from min to max in increments of 5 points:
  for (int fadeValue = 0 ; fadeValue <= 255; fadeValue += 5) {
    // sets the value (range from 0 to 255):
    for (int thispin = 0 ; thispin < 6; thispin ++)
    {
      analogWrite(_ledPinsPwm[thispin], fadeValue);
    }
    // wait for 100 milliseconds to see the dimming effect
    mydelay(_period_100);
  }

  // fade out from max to min in increments of 5 points:
  for (int fadeValue = 255 ; fadeValue >= 0; fadeValue -= 5) {
    // sets the value (range from 0 to 255):
    for (int thispin = 0 ; thispin < 6; thispin ++)
    {
      analogWrite(_ledPinsPwm[thispin], fadeValue);
    }
    // wait for 100 milliseconds to see the dimming effect
    mydelay(_period_100);
  }
}

//14
void Display_LED_lib::Ledmode_random()
{
  Ledmode_alloff();
  uint8_t ranNumpin = random(3, 18);
  digitalWrite(ranNumpin, HIGH);
  mydelay(_period_500);
}

//15
void Display_LED_lib::Ledmode_random_2()
{
  Ledmode_alloff();
  uint8_t ranNumpin_red = random(3, 8);
  uint8_t ranNumpin_amber = random(8, 13);
  uint8_t ranNumpin_green = random(13, 18);
  digitalWrite(ranNumpin_red, HIGH);
  digitalWrite(ranNumpin_amber, HIGH);
  digitalWrite(ranNumpin_green, HIGH);
  mydelay(_period_500);
}

//16
void Display_LED_lib::Ledmode_random_3()
{
  Ledmode_alloff();
  for (int count = 0; count <= 3; count++)
  {
    uint8_t ranNumpin_red = random(3, 8);
    uint8_t ranNumpin_amber = random(8, 13);
    uint8_t ranNumpin_green = random(13, 18);
    digitalWrite(ranNumpin_red, HIGH);
    digitalWrite(ranNumpin_amber, HIGH);
    digitalWrite(ranNumpin_green, HIGH);
  }
  mydelay(_period_1000);
}

//17
void Display_LED_lib::Ledmode_redone()
{
  Ledmode_alloff();
  mydelay(_period_1000);
  digitalWrite(_ledPinsRed[2], HIGH);
  mydelay(_period_1000);
}

//18
void Display_LED_lib::Ledmode_amberone()
{
  Ledmode_alloff();
  mydelay(_period_1000);
  digitalWrite(_ledPinsAmber[2], HIGH);
  mydelay(_period_1000);
}

//19
void Display_LED_lib::Ledmode_greenone()
{
  Ledmode_alloff();
  mydelay(_period_1000);
  digitalWrite(_ledPinsGreen[2], HIGH);
  mydelay(_period_1000);
}

//20
void Display_LED_lib::Ledmode_all()
{
  Ledmode_allon();
  Ledmode_redon();
  Ledmode_amberon();
  Ledmode_greenon();
  Ledmode_knightrider();
  Ledmode_knightrider_2();
  Ledmode_knightrider_3();
  Ledmode_knightrider_4();
  Ledmode_trafficlight();
  Ledmode_allblink();
  Ledmode_trafficlight_fast();
  Ledmode_pwm();
  Ledmode_random();
  Ledmode_random_2();
  Ledmode_random_3();
  Ledmode_redone();
  Ledmode_amberone();
  Ledmode_greenone();
}

//Method to replace delay with millis();
void mydelay(int period)
{
  unsigned long time_now = 0;
time_now = millis();
   while(millis() < time_now + period){
        //wait approx. [period] ms
      __asm__("nop\n\t");
    }
}